wayland-client/src/globals.rs

```
@@ -0,0 +1,291 @@
+    ///
+    /// This list is only valid until the next time the events are processed from the connection.
+    pub fn globals(&self) -> Vec<Global> {
+        let globals = self.registry.data::<Mutex<Vec<Global>>>().unwrap();
+        let guard = globals.lock().unwrap();
+        guard.clone()
+    }
+
+    /// Binds a global, returning a new protocol object associated with the global.
+    ///
+    /// The `version` specifies the range of versions that should be bound. This function will guarantee the
+    /// version of the returned protocol object is the lower of the maximum requested version and the advertised
+    /// version.
+    ///
+    /// If the lower bound of the `version` is less than the version advertised by the server, then
+    /// [`BindError::UnsupportedVersion`] is returned.
+    ///
+    /// ## Multi-instance/Device globals.
+    ///
+    /// This function is not intended to be used with globals that have multiple instances such as `wl_output`
+    /// and `wl_seat`. These types of globals need their own initialization mechanism because these
+    /// multi-instance globals may be removed at runtime. To handle then, you should instead rely on the
+    /// `Dispatch` implementation for `WlRegistry` of your `State`.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the maximum requested version is greater than the known maximum version of
+    /// the interface. The known maximum version is determined by the code generated using wayland-scanner.
+    pub fn bind<I, State, U>(
+        &self,
+        qh: &QueueHandle<State>,
+        version: RangeInclusive<u32>,
+        udata: U,
+    ) -> Result<I, BindError>
+    where
+        I: Proxy + 'static,
+        State: Dispatch<I, U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let version_start = *version.start();
+        let version_end = *version.end();
+        let interface = I::interface();
+
+        if *version.end() > interface.version {
```

I'd agree with the coverage warning and I do think we should test this does indeed panic given an invalid version. I guess bind wl_compositor with a version of `u32::MAX`